src/tools/webrtc-test-client/src/pybind.cpp

```diff
@@ -0,0 +1,134 @@
+     * https://docs.python.org/3/c-api/init.html#non-python-created-threads
+     */
+    std::cout<<"SDP offer: "<<offer<<std::endl;
+    gstate = PyGILState_Ensure();
+    offer_future.attr("set_result")(offer);
+    PyGILState_Release(gstate);
+}
+
+// void offer_callback_chayan(std::string offer) {
+//     /* This is being called from a thread which is not the main, read
+//      * https://docs.python.org/3/c-api/init.html#non-python-created-threads
+//      */
+//     std::cout<<"SDP offer: "<<offer<<std::endl;
+//     // gstate = PyGILState_Ensure();
+//     // offer_future.attr("set_result")(offer);
+//     // PyGILState_Release(gstate);
+// }
+
+void answer_callback(std::string answer) {
+    /* This is being called from a thread which is not the main, read
+     * https://docs.python.org/3/c-api/init.html#non-python-created-threads
+     */
+    gstate = PyGILState_Ensure();
+    answer_future.attr("set_result")(answer);
+    PyGILState_Release(gstate);
+}
+
+void create_offer(py::object future) {
+    /*
+     * We want this thread to release the gil so that python code keeps
+     * executing.
+     */
+    // gil = PyEval_SaveThread();
+    gstate = PyGILState_Ensure();
+    std::cout<<"gstate ensured in create_offer"<<std::endl;
+    offer_future = future;
+    std::cout<<"future initialised"<<std::endl;
+    pywebrtc::create_offer(offer_callback);
+    std::cout<<"pywebrtc::create_offer called"<<std::endl;
+    PyGILState_Release(gstate);
+    std::cout<<"gstate released in create_offer"<<std::endl;
+    // PyEval_RestoreThread(gil);
+}
+
+void create_offer_async(std::function<void(std::string)> offer_callback_async) {
+    /*
+     * We want this thread to release the gil so that python code keeps
+     * executing.
+     */
+    // gil = PyEval_SaveThread();
+    gstate = PyGILState_Ensure();
+    std::cout<<"gstate ensured in create_offer"<<std::endl;
+    
+    pywebrtc::create_offer(offer_callback_async);
+    std::cout<<"pywebrtc::create_offer called"<<std::endl;
+    PyGILState_Release(gstate);
+    std::cout<<"gstate released in create_offer"<<std::endl;
+    // PyEval_RestoreThread(gil);
+}
+
+void create_answer(std::string offer, py::object future) {
+    gil = PyEval_SaveThread();
+    answer_future = future;
+    pywebrtc::create_answer(offer, answer_callback);
+    PyEval_RestoreThread(gil);
+}
+
+void set_answer(const std::string& answer) {
+    gil = PyEval_SaveThread();
+    pywebrtc::set_answer(answer);
+    PyEval_RestoreThread(gil);
+}
+
+void constructor(std::function<void(std::string,std::string)> logging_callback, 
+                 std::function<void(std::string)> on_message_callback) {
+    python_logging = logging_callback;
+    python_on_message =  on_message_callback;
+    pywebrtc::set_logging(logging_callback);
+    pywebrtc::set_on_message(python_on_message);
+    pywebrtc::initialize();
+}
+
+auto cleanup_callback = []() {
+    pywebrtc::destructor();
+};
+
+
+PYBIND11_MODULE(pybind, m) {
```

Comment:
using pybind will make it harder for us to create testing packages because then the wheel will depend on the python binary version.

Could we use ffi instead like other python modules? Can this be placed into testing infrastructure and not as a stand-alone build?

Reply:
Agreed, we initially thought pybind was the commonly used bindings in Matter. We found out it was changed recently. We have started to modify it.